crypto/fipsmodule/ec/ec_key.c

```diff
@@ -562,3 +567,123 @@ void *EC_KEY_get_ex_data(const EC_KEY *d, int idx) {
+  }
+
+  if(eckey_meth) {
+    *ret = *eckey_meth;
+  }
+  return ret;
+}
+
+void EC_KEY_METHOD_free(EC_KEY_METHOD *eckey_meth) {
+  if(eckey_meth != NULL) {
+    OPENSSL_free(eckey_meth);
+  }
+}
+
+int EC_KEY_set_method(EC_KEY *ec, const EC_KEY_METHOD *meth) {
+  if(ec == NULL || meth == NULL) {
+    OPENSSL_PUT_ERROR(EC, ERR_R_PASSED_NULL_PARAMETER);
+    return 0;
+  }
+
+  // These fields are currently not supported by AWS-LC and cannot be set
+  if(meth->compute_key || meth->copy || meth->keygen || meth->set_group ||
+  meth->set_private || meth->set_public || meth->verify ||
+  meth->verify_sig || meth->sign_setup) {
+    OPENSSL_PUT_ERROR(EC, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+    return 0;
+  }
+
+  ec->eckey_method = meth;
+  return 1;
+}
+
+const EC_KEY_METHOD *EC_KEY_get_method(const EC_KEY *ec) {
+  if(ec == NULL) {
+    OPENSSL_PUT_ERROR(EC, ERR_R_PASSED_NULL_PARAMETER);
+    return NULL;
+  }
+
+  return ec->eckey_method;
+}
+
+void EC_KEY_METHOD_set_init(EC_KEY_METHOD *meth,
+                            int (*init)(EC_KEY *key),
+                            void (*finish)(EC_KEY *key),
+                            int (*copy)(EC_KEY *dest, const EC_KEY *src),
+                            int (*set_group)(EC_KEY *key, const EC_GROUP *grp),
+                            int (*set_private)(EC_KEY *key,
+                                               const BIGNUM *priv_key),
+                            int (*set_public)(EC_KEY *key,
+                                              const EC_POINT *pub_key)) {
+  if(meth == NULL) {
+    OPENSSL_PUT_ERROR(EC, ERR_R_PASSED_NULL_PARAMETER);
+    return;
+  }
+
+  // Setting these fields is currently not supported by AWS-LC
+  assert(!copy && !set_group && !set_private && !set_public);
+  if(copy || set_group || set_private || set_public) {
+    OPENSSL_PUT_ERROR(EC, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+    abort();
+  }
+
+  meth->init = init;
+  meth->finish = finish;
+}
+
+void EC_KEY_METHOD_set_sign(EC_KEY_METHOD *meth,
+                            int (*sign)(int type, const uint8_t *digest,
+                                    int digest_len, uint8_t *sig,
+                                    unsigned int *siglen, const BIGNUM *k_inv,
+                                    const BIGNUM *r, EC_KEY *eckey),
+                            int (*sign_setup)(EC_KEY *eckey, BN_CTX *ctx_in,
+                                              BIGNUM **kinvp, BIGNUM **rp),
+                            ECDSA_SIG *(*sign_sig)(const uint8_t *digest,
+                                    int digest_len,
+                                    const BIGNUM *in_kinv, const BIGNUM *in_r,
+                                    EC_KEY *eckey)) {
+
+  if(meth == NULL) {
+    OPENSSL_PUT_ERROR(EC, ERR_R_PASSED_NULL_PARAMETER);
+    return;
+  }
+
+  // Setting this field is currently not supported by AWS-LC
+  if(sign_setup) {
+    OPENSSL_PUT_ERROR(EC, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+    abort();
+  }
```

Comment:
- Discussed offline. Static asserts would not work to determine the nullness of a runtime value. Putting an error on the queue and returning could lead to erroneous usage if the error queue is not checked. Aborting, while abrupt, is documented and will provide a clearer indication of unsupported functionality. 
- We define some _NOT_SUPPORTED_ ERR types but they are library specific (EVP, DH, [ENGINE](https://github.com/smittals2/aws-lc/blob/3c7ba297d6150cbe65b636c507e97abeeb65822f/include/openssl/engine.h#L114), etc.). We could define one for this use case if needed.

Reply:
Using the static asserts in a header file to prevent compilation with consumers that set specific fields to Non-NULL values worked. PR has been updated with relevant changes.